opal/mca/common/ofi/common_ofi.c

```diff
@@ -510,177 +522,220 @@ static int compute_dev_distances(pmix_device_distance_t **distances,
-    int ret, idx = 0;
+    int ret;
     pmix_data_array_t *dptr;
-    uint16_t near = USHRT_MAX;
+    pmix_device_distance_t *distances;
     pmix_info_t directive;
     pmix_value_t *val = NULL;
-    pmix_device_distance_t *distances, *nearest = NULL;
+    size_t ndist, num_nearest = 0;
+    struct fi_info *current_provider = NULL;
+    uint16_t dists[num_providers], *dist = NULL, min_dist = USHRT_MAX;
+    uint32_t provider_rank = 0;
 
     PMIx_Info_load(&directive, PMIX_OPTIONAL, NULL, PMIX_BOOL);
-    ret = PMIx_Get(&opal_process_info.myprocid,
-             PMIX_DEVICE_DISTANCES, &directive, 1, &val);
+    ret = PMIx_Get(&opal_process_info.myprocid, PMIX_DEVICE_DISTANCES, &directive, 1, &val);
     PMIx_Info_destruct(&directive);
     if (ret != PMIX_SUCCESS || !val) {
         ret = compute_dev_distances(&distances, &ndist);
         if (ret) {
+            ret = OPAL_ERROR;
             goto out;
         }
         goto find_nearest;
     }
 
     if (PMIX_DATA_ARRAY != val->type) {
+        ret = OPAL_ERROR;
         goto out;
     }
     dptr = val->data.darray;
     if (NULL == dptr) {
+        ret = OPAL_ERROR;
         goto out;
     }
     if (PMIX_DEVICE_DIST != dptr->type) {
+        ret = OPAL_ERROR;
         goto out;
     }
 
-    distances = (pmix_device_distance_t*)dptr->array;
+    distances = (pmix_device_distance_t *) dptr->array;
     ndist = dptr->size;
 
 find_nearest:
-    nearest = calloc(sizeof(*distances), ndist);
-    if (!nearest) {
-        goto out;
-    }
-
-    for (i = 0; i < ndist; i++) {
-        if (distances[i].type != PMIX_DEVTYPE_NETWORK &&
-            distances[i].type != PMIX_DEVTYPE_OPENFABRICS)
+    for (current_provider = provider_list, dist = dists; NULL != current_provider;
+         current_provider = current_provider->next, ++dist) {
+        if (OPAL_SUCCESS != check_provider_attr(provider_list, current_provider)) {
             continue;
-        if (distances[i].mindist < near) {
-            idx = 0;
-            near = distances[i].mindist;
-            nearest[idx] = distances[i];
-            idx++;
-        } else if (distances[i].mindist == near) {
-            nearest[idx] = distances[i];
-            idx++;
+        }
+        if (OPAL_SUCCESS != get_provider_distance(topology, current_provider, distances, ndist, dist)) {
+            *dist = USHRT_MAX;
+        }
+
+        if (*dist < min_dist) {
+            min_dist = *dist;
+            num_nearest = 1;
+        } else if (*dist == min_dist) {
+            ++num_nearest;
+        }
+
+        if (OPAL_SUCCESS == check_provider_attr(provider_list, current_provider)) {
+            opal_output_verbose(1, opal_common_ofi.output, "provider: %s dist: %d",
+                                current_provider->domain_attr->name, *dist);
         }
     }
 
-    *num_distances = idx;
+    ret = OPAL_ERROR;
+    if (0 >= num_nearest) {
+        return ret;
+    }
 
+    provider_rank = rank % num_nearest;
+    num_nearest = 0;
```

Comment:
I noticed a little nuance between `num_nearest++` vs `--num_nearest`.

`--num_nearest` will result in `NIC-3` being selected before `NIC-2`, i.e. `NIC-3`, `NIC-2`, `NIC-1`, `NIC-0`.

This is counter intuitive - therefore I inverted the order so that `NIC-0` is selected before `NIC-1`, `NIC-2`, `NIC-3`. Just a little nicer to human brain.